assets/js/modules/analytics-4/components/common/WebDataStreamNameInput.js

```diff
@@ -0,0 +1,150 @@
+import { __ } from '@wordpress/i18n';
+import { isURL } from '@wordpress/url';
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import { CORE_FORMS } from '../../../../googlesitekit/datastore/forms/constants';
+import { CORE_SITE } from '../../../../googlesitekit/datastore/site/constants';
+import {
+	FORM_SETUP,
+	MODULES_ANALYTICS_4,
+	WEBDATASTREAM_CREATE,
+} from '../../datastore/constants';
+import { TextField } from 'googlesitekit-components';
+import VisuallyHidden from '../../../../components/VisuallyHidden';
+import WarningIcon from '../../../../../svg/icons/warning-v2.svg';
+import {
+	isValidPropertyID,
+	isValidWebDataStreamName,
+} from '../../utils/validation';
+
+const { useSelect, useDispatch } = Data;
+
+export default function WebDataStreamNameInput() {
+	const propertyID = useSelect( ( select ) =>
+		select( MODULES_ANALYTICS_4 ).getPropertyID()
+	);
+	const webDataStreamID = useSelect( ( select ) =>
+		select( MODULES_ANALYTICS_4 ).getWebDataStreamID()
+	);
+	const webDataStreamName = useSelect( ( select ) =>
+		select( CORE_FORMS ).getValue( FORM_SETUP, 'webDataStreamName' )
+	);
+	const webDataStreamAlreadyExists = useSelect( ( select ) =>
+		isValidPropertyID( propertyID )
+			? select( MODULES_ANALYTICS_4 ).doesWebDataStreamExist(
+					propertyID,
+					webDataStreamName
+			  )
+			: false
+	);
+	const siteURL = useSelect( ( select ) =>
+		select( CORE_SITE ).getReferenceSiteURL()
+	);
+
+	const { setValues } = useDispatch( CORE_FORMS );
+
+	const onChange = useCallback(
+		( { currentTarget } ) => {
+			setValues( FORM_SETUP, { webDataStreamName: currentTarget.value } );
+		},
+		[ setValues ]
+	);
+
+	// Set the default web data stream name.
+	useMount( () => {
+		if ( ! webDataStreamName && isURL( siteURL ) ) {
+			const { hostname } = new URL( siteURL );
+
+			setValues( FORM_SETUP, {
+				webDataStreamName: hostname,
+			} );
+		}
+	} );
+
+	// Bounce if an existing web data stream is selected.
+	if ( webDataStreamID !== WEBDATASTREAM_CREATE ) {
+		return null;
+	}
+
+	const error =
+		webDataStreamAlreadyExists ||
+		! webDataStreamName ||
+		! isValidWebDataStreamName( webDataStreamName );
+
+	return (
+		<div className="googlesitekit-analytics-webdatastreamname">
+			<TextField
+				className={ classnames( {
+					'mdc-text-field--error': error,
+				} ) }
+				label={ __( 'Web Data Stream Name', 'google-site-kit' ) }
+				outlined
+				helperText={ ( () => {
+					if ( webDataStreamAlreadyExists ) {
+						return __(
+							'A web data stream with this name already exists.',
+							'google-site-kit'
+						);
+					} else if ( ! webDataStreamName ) {
+						return __(
+							'A web data stream name is required.',
+							'google-site-kit'
+						);
+					} else if (
+						! isValidWebDataStreamName( webDataStreamName )
+					) {
+						return __(
+							'This is not a valid web data stream name.',
+							'google-site-kit'
+						);
+					}
+
+					return false;
+				} )() }
```

Comment:
Instead of inlining this function and immediately invoking, maybe move it up to the component body. Otherwise, in the first glance, it seems like `helperText` actually supports passing a function. :smile:

Reply:
Updated, thanks!